src/python/base/utils.py

```diff
@@ -782,16 +787,20 @@ def wait_until_timeout(threads, thread_timeout):
 
 def write_data_to_file(content, file_path, append=False):
   """Writes data to file."""
-  content_string = str(content)
   failure_wait_interval = environment.get_value('FAIL_WAIT')
   file_mode = 'ab' if append else 'wb'
   retry_limit = environment.get_value('FAIL_RETRIES')
 
-  for _ in range(retry_limit):
+  # TODO(mbarbella): Require callers to pass strings or bytes then reduce this
+  # back to the original retry limit.
+  for _ in range(retry_limit + 1):
     try:
       with open(file_path, file_mode) as file_handle:
-        file_handle.write(content_string)
-    except:
+        file_handle.write(content)
+    except TypeError:
+      content = str(content)
```

Comment:
can we just always try convert it to the right type?

also, file_mode is 'ab' or 'wb'. Wouldn't writing a str type here fail on python 3 ?

```python
>>> with open('test', 'wb') as f:
...   f.write('123')
...
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
TypeError: a bytes-like object is required, not 'str'
```

Reply:
Good catch on the type issue, I'll fix that.

As for always converting, we have a few cases where we depend on the Python 2 behavior for some of this. For example we might pass the integer 0 to this function and expect to write "0" to the file (converting to bytes in Python 3 will write an empty file).

I think the problem is really that we're just expecting this function to work a bit too magically. For now I think trying to catch the type error here is the right way to get this working with the behavior we expect, but later callers should just pass sane input to this function.